src/cpp/subscriber/SubscriberHistory.cpp

```diff
@@ -33,23 +33,37 @@ namespace fastrtps {
 
 using namespace rtps;
 
+static HistoryAttributes to_history_attributes(
+        const HistoryQosPolicy& history,
+        const ResourceLimitsQosPolicy& resource,
+        SubscriberImpl* simpl,
+        uint32_t payloadMaxSize,
+        MemoryManagementPolicy_t mempolicy)
+{
+    auto initial_samples = resource.allocated_samples;
+    auto max_samples = resource.max_samples;
+
+    if (history.kind != KEEP_ALL_HISTORY_QOS)
+    {
+        max_samples = history.depth;
+        if (simpl->getAttributes().topic.getTopicKind() != NO_KEY)
+        {
+            max_samples *= resource.max_instances;
+        }
+
+        initial_samples = std::min(initial_samples, max_samples);
+    }
+
+    return HistoryAttributes(mempolicy, payloadMaxSize, initial_samples, max_samples);
+}
+
 SubscriberHistory::SubscriberHistory(
         SubscriberImpl* simpl,
         uint32_t payloadMaxSize,
         const HistoryQosPolicy& history,
         const ResourceLimitsQosPolicy& resource,
         MemoryManagementPolicy_t mempolicy)
-    : ReaderHistory(HistoryAttributes(mempolicy, payloadMaxSize,
-                history.kind == KEEP_ALL_HISTORY_QOS ?
-                        resource.allocated_samples :
-                        simpl->getAttributes().topic.getTopicKind() == NO_KEY ?
-                            std::min(resource.allocated_samples, history.depth) :
-                            std::min(resource.allocated_samples, history.depth * resource.max_instances),
-                history.kind == KEEP_ALL_HISTORY_QOS ?
-                        resource.max_samples :
-                        simpl->getAttributes().topic.getTopicKind() == NO_KEY ?
-                            history.depth :
-                            history.depth * resource.max_instances))
+    : ReaderHistory(to_history_attributes(history, resource, simpl, payloadMaxSize, mempolicy))
```

Comment:
We could do the following, and keep `to_history_attributes` exactly as in #1314
```suggestion
    : ReaderHistory(to_history_attributes(simpl->getAttributes().topic, payloadMaxSize, mempolicy))
```